src/Microsoft.AspNetCore.WebSockets/WebSocketOptions.cs

```diff
@@ -27,5 +29,11 @@ public WebSocketOptions()
         /// The default is 4kb.
         /// </summary>
         public int ReceiveBufferSize { get; set; }
+
+        /// <summary>
+        /// Set the Origin header values allowed for WebSocket requests to prevent Cross-Site WebSocket Hijacking.
+        /// By default all Origins are allowed.
+        /// </summary>
+        public IList<string> AllowedOrigins { get; }
```

Comment:
Not pleased with the duplication between WebSocketOptions.AllowedOrigins and CorsPolicy.Origins. CorsPolicy provides much richer functionality that we should be taking advantage of here.

Who's your PM?

Reply:
Not me :) But yes, using the existing classes would be better.

Reply:
I'm not yet clear on why CORS is applicable here. WebSockets is not covered by the CORS spec, this is based on the [Origin validation](https://tools.ietf.org/html/rfc6455#section-10.2) specified in the WebSocket spec.

Do we need to add Header validation? And if so do we just allow the required WebSocket handshake headers? Method validation (WebSockets only uses `GET`)?

I'm on board with reducing duplication here if it's applicable, but I'm trying to understand exactly what from CORS applies here when it's not part of the scope of CORS. I'm hesitant to add a bunch of out-of-spec logic because that means we're inventing new policies which requires a bunch more logic.

If you're talking about just sharing the logic for matching origins via shared-source, I'm certainly on-board there.

> Who's your PM?

Not a clue. This falls under SignalR, and I've never been clear on who exactly the PM is :). @DamianEdwards ?

Reply:
Interesting, I hadn't realized this was independent of the CORS spec. The Origin and functional overlap make them seem like the same feature set.
We're still duplicating options and code that we'd rather not, but I can see why the full CORS policy infrastructure might not be appropriate here.

Reply:
Some of that was slightly confused by @BrennanConroy putting "CORS" in the PR title 😛 . It's CORS-esque, but it's not part of CORS. It's a separate concern described in the WebSocket spec.

Do agree with getting the logic shared, but there's not much of it. The "shared" logic really just boils down to comparing the list of allowed origins with the actual origin value, and that seems to be a little more complicated in the CorsPolicy logic. Is it worth getting that into a shared-source package? I'm not sure it is.

Reply:
I'm more concerned about the user not having to configure their origins list in two places.

Reply:
Well, but CORS has multiple policies, so it's not as simple as a single list. Should we support providing a CORS policy and only use the Origins from it? That seems a little more confusing to a user to me given that we're _partially_ applying the policy they provide. I'm not entirely opposed though, if we can find a clean way to do it.

@BrennanConroy It looks like `CorsMiddleware` already handles the issue you mentioned earlier where it takes an options object in the `Use` but still manages to be activated via DI.